Grant a condition while performing an external capture

gwenzek opened the pull request:
Add a notifier for the beginning of an external capture
Implement uncloaking when a capture begins (#13767)

The attached replay demonstrate engineers modded with the cloaking trait that get uncloaked when they starts capturing an enemy building.
[OpenRA-2017-08-15T165036Z.orarep.zip](https://github.com/OpenRA/OpenRA/files/1225786/OpenRA-2017-08-15T165036Z.orarep.zip)

I'm not sure what will happen if the cloaked unit finish capturing the building.
Should it get cloaked again ? Does that happen automatically ? Could that happen while the unit is still capturing the building ?

MustaphaTR commented:
Engineer is cloaking back while it is still capturing with this PR. I checked Generals to be sure, at there Infantries are uncloaked during the whole capturing.

gwenzek commented:
Thanks for the review.
I changed so the unit doesn't cloak while capturing. 
I added a "OnCaptureCancelled" event to detect when the capture is cancelled so the unit will cloak back.
I haven't see other "ActionCancelled" events so I'm not sure that's the correct pattern.

MustaphaTR reviewed (approved):
👍

SoScared commented:
Just to be clear, does this add a notifier with engineer capturing in RA?

MustaphaTR commented:
No. This adds a Capturing, cloak RevealOn type, so i can make my GLA Rebels with Camouflage reveal while capturing on my mod.

gwenzek commented:
@SoScared Yes my implementation adds a notifier for engineer capturing in RA that's how I tested it.
The traits Cloak listen for this notifier and uncloack the unit when capturing.

Smittytron commented:
> Yes my implementation adds a notifier for engineer capturing in RA that's how I tested it.
> The traits Cloak listen for this notifier and uncloack the unit when capturing.

But not a notification that players hear right? Asking because a few weeks ago someone was suggesting that you should be notified when your building is being captured. This suggestion is universally opposed by the competitive community.

MustaphaTR commented:
No, this is not a notification. A notifier and notification is different things. I think notification was what SoScared meant too, thats why i answered no. A notifier allows stuff a trait does to be used by another trait, in this case making ExternalCaptures uncloak a unit.

pchote commented on OpenRA.Mods.Common/Activities/ExternalCaptureActor.cs:
```diff
@@ -49,8 +50,11 @@ public override Activity Tick(Actor self)
 				return ActivityUtils.SequenceActivities(new MoveAdjacentTo(self, target), this);
 
-			if (!capturable.CaptureInProgress)
+			if (!capturable.CaptureInProgress) {
```
Please note that the openra [coding standard](https://github.com/OpenRA/OpenRA/wiki/Coding-Standard) has braces on their own lines.  This applies to a few places in this PR.

pchote commented on OpenRA.Mods.Common/Traits/Cloak.cs:
```diff
@@ -212,5 +214,22 @@ void INotifyHarvesterAction.Undocked()
 			isDocking = false;
 		}
+
+		void INotifyExternalCapture.OnCapturing(Actor self, Actor captor, Player oldOwner, Player newOwner)
```
It would be a better fit to have `ExternalCapturable` grant a condition that modders can choose to use to disable this (or any other) trait rather than hardcoding the behaviour in here explicitly.

This is more in line with how how we handle damage states (using conditions) vs one-off notifications, which are mostly hardcoded.

gwenzek commented:
Thanks @MustaphaTR for the clarifications. I didn't properly understand SoScared concerns.

@pchote Thanks for your suggestions. This is my first PR for OpenRA and wasn't aware of this Condition granting mechanism. It look interesting. I just have small concerns about using it here.

Currently the YAML for a cloaking unit looks like:
```
	Cloak:
		InitialDelay: 250
		CloakDelay: 120
		UncloakOn: Attack, Unload, Infiltrate, Demolish
		CloakTypes: Cloak
		IsPlayerPalette: true
```

My PR would allow to just add "Capturing" to the "UncloakOn" property list.
Using a condition, it will look something like:

```
	Cloak:
		InitialDelay: 250
		CloakDelay: 120
		UncloakOn: Attack, Unload, Infiltrate, Demolish
		CloakTypes: Cloak
		IsPlayerPalette: true
      		RequiresCondition: !capturing
```

Don't you find that a bit awkward to have to separate settings for controlling the Uncloaking ?

gwenzek committed:
add a "capturing" condition for ExternalCapture

gwenzek commented:
I've started an implementation using the Trait Conditions.
The code is much simpler.
If you agree that it's not a problem to use Trait Conditions while other uncloaking events happen in another fashion, I'll create a new PR.
Meanwhile it's available here: https://github.com/OpenRA/OpenRA/compare/bleed...gwenzek:uncloak_on_capture

pchote commented:
> Don't you find that a bit awkward to have two separate settings for controlling the Uncloaking ?

Nope, we already do exactly this to disable cloak on critical health in RA/TD/TS.  See this example from the TD stealth tank:
https://github.com/OpenRA/OpenRA/blob/898ef8fe50280847a518bc28c469140b14d99763/mods/cnc/rules/vehicles.yaml#L548-L556

MustaphaTR commented:
Currently disabling capturing while unit is damaged to red is handled by conditions too. I think it would be better to use conditions,

gwenzek commented on OpenRA.Mods.Common/Traits/Cloak.cs:
```diff
@@ -135,7 +135,11 @@ IEnumerable<IRenderable> IRenderModifier.ModifyRender(Actor self, WorldRenderer
-			if (!IsTraitDisabled)
+			if (IsTraitDisabled)
+			{
+				Uncloak();
```
When the trait get disabled and re-enabled we want the cloaking timer to be resetted.

pchote replied:
This will set the countdown timer to `Info.CloakDelay`, which will override the `InitialDelay` value that we want if the trait starts off disabled (e.g. being enabled on veterancy or from a crate).  It would be better to implement and call this from `TraitDisabled`, which only runs when the trait goes from enabled &rarr; disabled.

gwenzek replied:
done

gwenzek commented on OpenRA.Mods.Common/Traits/ExternalCaptures.cs:
```diff
@@ -34,6 +34,10 @@ class ExternalCapturesInfo : ITraitInfo
 
+		[GrantedConditionReference]
+		[Desc("Condition granted when capturing.")]
+		public readonly string CapturingCondition = "capturing";
```
Shall I set a default value here ? What's the point of forcing the modder to set it ?

pchote replied:
No, please default this to nothing.  Forcing the modder to enable it prevents bugs when names implicitly overlap, and helps a tiny bit with performance (no point paying the cost of managing conditions that are never used).

pchote commented on OpenRA.Mods.Common/Activities/ExternalCaptureActor.cs:
```diff
@@ -76,6 +79,7 @@ public override Activity Tick(Actor self)
 							t.OnCapture(target.Actor, self, oldOwner, self.Owner);
 
 						capturable.EndCapture();
```
This is done from inside `EndCapture`, so can be removed from here.

gwenzek replied:
done

pchote commented on OpenRA.Mods.Common/Activities/ExternalCaptureActor.cs:
```diff
@@ -38,9 +41,7 @@ public override Activity Tick(Actor self)
-				if (capturable.CaptureInProgress)
-					capturable.EndCapture();
-
+				EndCapture(self);
```
You need to call this from the case above too, otherwise the condition will remain permanently granted if the target actor dies while being captured.

gwenzek replied:
good catch ! Done

pchote reviewed (commented):
Looking good overall.  Just two issues, then I expect to :+1:.

pchote commented:
@MustaphaTR: This has changed a lot since your original :+1:, so can you please check this again once the fixups are done?

gwenzek commented:
Fixup done.

Also what's the best way to notify the reviewers I fixed the issue they spotted and they can re-review the code ? Do I need to write in this thread or just answering in the diff view works ?

pchote commented:
Answering in the diff view works, but it can sometimes (unfortunately with our current lack of reviewing bandwidth, that has regressed to "usually") take a while for someone to come back to it.

Sorry! and thanks for the reminder - I'll try and recheck this tonight.

abcdefg30 commented on OpenRA.Mods.Common/Activities/ExternalCaptureActor.cs:
```diff
@@ -50,7 +48,9 @@ public override Activity Tick(Actor self)
 
 			if (!capturable.CaptureInProgress)
-				capturable.BeginCapture(self);
+			{
```
Braces around single statements should be removed according to https://github.com/OpenRA/OpenRA/wiki/Coding-Standard.

gwenzek replied:
yep sorry about that, I thought that the `if` and `else` should be braced similarly but no.

pchote commented on OpenRA.Mods.Common/Activities/ExternalCaptureActor.cs:
```diff
@@ -92,5 +90,20 @@ public override Activity Tick(Actor self)
+
+		void EndCapture(Actor self)
+		{
+			if (capturable.CaptureInProgress)
```
This needs to check `target.Type == TargetType.Actor` too, otherwise `capturable.EndCapture` will run on dead actors which then crashes the game with:
```
Exception of type `System.InvalidOperationException`: Attempted to get trait from destroyed object (proc 175 (not in world))
  at OpenRA.TraitDictionary.CheckDestroyed (OpenRA.Actor actor) [0x00026] in /Users/paul/src/OpenRA/OpenRA.Game//TraitDictionary.cs:84 
  at OpenRA.TraitDictionary.GetOrDefault[T] (OpenRA.Actor actor) [0x00002] in /Users/paul/src/OpenRA/OpenRA.Game//TraitDictionary.cs:94 
  at OpenRA.Actor.TraitOrDefault[T] () [0x0000d] in /Users/paul/src/OpenRA/OpenRA.Game//Actor.cs:248 
  at OpenRA.Mods.Common.Traits.ExternalCapturable.EndCapture () [0x00007] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//Traits/ExternalCapturable.cs:80 
  at OpenRA.Mods.Common.Activities.ExternalCaptureActor.EndCapture (OpenRA.Actor self) [0x00017] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//Activities/ExternalCaptureActor.cs:104 
  at OpenRA.Mods.Common.Activities.ExternalCaptureActor.Tick (OpenRA.Actor self) [0x00075] in /Users/paul/src/OpenRA/OpenRA.Mods.Common//Activities/ExternalCaptureActor.cs:41 
  at OpenRA.Activities.Activity.TickOuter (OpenRA.Actor self) [0x00061] in /Users/paul/src/OpenRA/OpenRA.Game//Activities/Activity.cs:175 
  at OpenRA.Traits.ActivityUtils.RunActivity (OpenRA.Actor self, OpenRA.Activities.Activity act) [0x00024] in /Users/paul/src/OpenRA/OpenRA.Game//Traits/ActivityUtils.cs:37 
  at OpenRA.Actor.Tick () [0x00010] in /Users/paul/src/OpenRA/OpenRA.Game//Actor.cs:163 
  at OpenRA.World.Tick () [0x000c3] in /Users/paul/src/OpenRA/OpenRA.Game//World.cs:345 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x00210] in /Users/paul/src/OpenRA/OpenRA.Game//Game.cs:604 
  at OpenRA.Game.LogicTick () [0x00055] in /Users/paul/src/OpenRA/OpenRA.Game//Game.cs:628 
  at OpenRA.Game.Loop () [0x000d6] in /Users/paul/src/OpenRA/OpenRA.Game//Game.cs:755 
  at OpenRA.Game.Run () [0x00042] in /Users/paul/src/OpenRA/OpenRA.Game//Game.cs:795 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in /Users/paul/src/OpenRA/OpenRA.Game//Support/Program.cs:136 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in /Users/paul/src/OpenRA/OpenRA.Game//Support/Program.cs:40 
```

gwenzek replied:
thanks, I added a test.
But how did you managed to crash the game ?
I tried to kill the captor and the target during the capture but nothing weird happened in both cases.

pchote replied:
For my tests I added `Inherits@foo: ^IronCurtainable` and `CapturingCondition: invulnerability` to `E6` in RA. I then tried out various situations ingame.  The crash occurs when the actor being captured (e.g. a power plant) is killed (e.g. using Tanya to demolish it) while the capture is underway.

pchote reviewed (commented):
:+1: after one last bug fix:

gwenzek committed:
check if target is alive

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9d437751454cc0937c4504611c71f8095388605c...66328cc1d2e699039f7e37412e5136dbe81808c7)
